server/gcpcredential/validate_test.go

```diff
@@ -0,0 +1,300 @@
+package gcpcredential
+
+import (
+	"bytes"
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"testing"
+	"time"
+
+	jwt "github.com/golang-jwt/jwt/v5"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/api/idtoken"
+)
+
+const testAudience = "testaud"
+const testKeyID = "testkid"
+
+func TestGoogleCACerts(t *testing.T) {
+	if _, err := googleCACerts(); err != nil {
+		t.Errorf("GoogleCACerts() returned error %v", nil)
+	}
+}
+
+type jwkInfo = map[string]any
+
+// Generates and returns an RSA256 private key and associated JWK.
+func testRSASigner(t *testing.T, keyID string) (*rsa.PrivateKey, JWK) {
+	t.Helper()
+
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatalf("Error generating test key: %v", err)
+	}
+
+	jwk := JWK{
+		Alg: "RS256",
+		Kid: keyID,
+		N:   base64.RawURLEncoding.EncodeToString(key.N.Bytes()),
+		E:   base64.RawURLEncoding.EncodeToString(big.NewInt(int64(key.E)).Bytes()),
+	}
+
+	return key, jwk
+}
+
+// Generates and returns a JWT token with the provided claims, audience, and signed by the provided key.
+func testGCPCredential(t *testing.T, claims *emailClaims, aud, keyID string, signer *rsa.PrivateKey) string {
+	t.Helper()
+
+	now := time.Now().Unix()
+
+	jwtClaims := jwt.MapClaims{
+		"aud":            aud,
+		"iss":            "accounts.google.com",
+		"exp":            now + 60,
+		"email":          claims.Email,
+		"email_verified": claims.EmailVerified,
+	}
+
+	token := jwt.NewWithClaims(jwt.SigningMethodRS256, jwtClaims)
+	token.Header = map[string]any{
+		"kid": keyID,
+		"alg": "RS256",
+	}
+	tokenString, err := token.SignedString(signer)
+	if err != nil {
+		t.Fatalf("Error generating token for %v: %v", claims, err)
+	}
+
+	return tokenString
+}
+
+type jwkFetcher struct {
+	jwkFunc func(req *http.Request) *http.Response
+}
+
+func (t *jwkFetcher) RoundTrip(req *http.Request) (*http.Response, error) {
+	return t.jwkFunc(req), nil
+}
+
+func testJWKFetcher(t *testing.T, jwks *JWKS) *jwkFetcher {
```

Comment:
nit: since jwks is the only meaningful parameter, this might read better as a field of jwkFetcher and you instantiate a new jwkFetcher with the actual JWKS in every test.